File: tests/oracle/typeop_local_type_1204.cc
/*
 * TYPEOP-LOCALTYPE-DISPATCH-0001: locked Ghidra 12.0.4 oracle projection for
 * TypeOpCall::getInputLocal.
 *
 * This fixture constructs one real FuncCallSpecs object, encodes it through
 * Funcdata::newVarnodeCallSpecs, and invokes PcodeOp::inputTypeLocal so the
 * unmodified Architecture-owned TypeOpCall is the code under observation.
 * It records FSPEC pointer aliasing, the same numeric offset in a non-FSPEC
 * space, slot-to-parameter indexing, per-parameter lock/this flags, size and
 * void guards, and canonical TypeFactory pointer identity.
 */
#include "bfd_arch.hh"
#include "funcdata.hh"
#include "libdecomp.hh"
#include "typeop.hh"

#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using namespace ghidra;
using namespace std;

string typeToken(const Datatype *ct)

{
  ostringstream stream;
  ct->printRaw(stream);
  return stream.str();
}

void emitBool(const string &key,bool value)

{
  cout << key << '=' << (value ? 1 : 0) << '\n';
}

void emitCase(const string &name,PcodeOp *op,int4 slot,Datatype *expected,
              ProtoParameter *param,TypeFactory *factory)

{
  Datatype *actual = op->inputTypeLocal(slot);
  Datatype *repeat = op->inputTypeLocal(slot);
  Datatype *fallback = factory->getBase(op->getIn(slot)->getSize(),TYPE_UNKNOWN);
  cout << "case." << name << ".slot=" << slot << '\n';
  cout << "case." << name << ".input_size=" << op->getIn(slot)->getSize() << '\n';
  cout << "case." << name << ".param_present=" << (param != (ProtoParameter *)0 ? 1 : 0) << '\n';
  cout << "case." << name << ".param_locked="
       << (param != (ProtoParameter *)0 && param->isTypeLocked() ? 1 : 0) << '\n';
  cout << "case." << name << ".param_this="
       << (param != (ProtoParameter *)0 && param->isThisPointer() ? 1 : 0) << '\n';
  cout << "case." << name << ".result_type=" << typeToken(actual) << '\n';
  cout << "case." << name << ".result_meta=" << static_cast<int4>(actual->getMetatype()) << '\n';
  cout << "case." << name << ".result_size=" << actual->getSize() << '\n';
  emitBool("case." + name + ".expected_identity",actual == expected);
  emitBool("case." + name + ".param_identity",
      param != (ProtoParameter *)0 && actual == param->getType());
  emitBool("case." + name + ".fallback_identity",actual == fallback);
  emitBool("case." + name + ".repeat_identity",actual == repeat);
}

ParameterPieces pieces(AddrSpace *space,uintb offset,Datatype *type,uint4 flags)

{
  ParameterPieces result;
  result.addr = Address(space,offset);
  result.type = type;
  result.flags = flags;
  return result;
}

void runFixture(const string &specDirectory,const string &binary)

{
  vector<string> specPaths(1,specDirectory);
  startDecompilerLibrary(specPaths);
  BfdArchitecture architecture(binary,"default",&cerr);
  DocumentStorage store;
  architecture.init(store);
  if (architecture.archid != "x86:LE:64:default:gcc")
    throw runtime_error("runtime architecture/compiler drifted: " + architecture.archid);

  TypeFactory *factory = architecture.types;
  AddrSpace *code = architecture.getDefaultCodeSpace();
  AddrSpace *reg = architecture.getSpaceByName("register");
  if (factory == (TypeFactory *)0 || code == (AddrSpace *)0 || reg == (AddrSpace *)0)
    throw runtime_error("required architecture service missing");

  Scope *global = architecture.symboltab->getGlobalScope();
  FunctionSymbol *symbol = global->addFunction(
      Address(code,0x500000),"typeop_call_local_fixture");
  Funcdata *fd = symbol->getFunction();

  const int4 pointerSize = factory->getSizeOfPointer();
  const int4 wordSize = architecture.getDefaultDataSpace()->getWordSize();
  Datatype *voidType = factory->getTypeVoid();
  Datatype *charType = factory->getTypeChar(factory->getSizeOfChar());
  Datatype *charPointer = factory->getTypePointer(pointerSize,charType,wordSize);
  Datatype *int4Type = factory->getBase(4,TYPE_INT);
  Datatype *uint8Type = factory->getBase(8,TYPE_UINT);
  Datatype *oversizeType = factory->getTypeArray(2,uint8Type);
  TypeStruct *objectType = factory->getTypeStruct("typeop_call_fixture_object");
  Datatype *objectPointer = factory->getTypePointer(pointerSize,objectType,wordSize);
  Datatype *plainPointer = factory->getTypePointer(pointerSize,int4Type,wordSize);

  PcodeOp *callOp = fd->newOp(9,Address(code,0x500010));
  fd->opSetOpcode(callOp,CPUI_CALL);
  // FuncCallSpecs::FuncCallSpecs reads the original direct-call target from
  // input 0 before flow replaces that input with the FSPEC annotation.
  fd->opSetInput(callOp,fd->newCodeRef(Address(code,0x600000)),0);
  FuncCallSpecs callspec(callOp);
  callspec.setInternal(architecture.defaultfp,voidType);
  callspec.setParam(0,"locked_ptr",
      pieces(reg,0x00,charPointer,ParameterPieces::typelock));
  callspec.setParam(1,"locked_small",
      pieces(reg,0x08,int4Type,ParameterPieces::typelock));
  callspec.setParam(2,"unlocked_int",
      pieces(reg,0x10,int4Type,0));
  callspec.setParam(3,"locked_void",
      pieces(reg,0x18,voidType,ParameterPieces::typelock));
  callspec.setParam(4,"locked_oversize",
      pieces(reg,0x20,oversizeType,ParameterPieces::typelock));
  callspec.setParam(5,"this_struct",
      pieces(reg,0x28,objectPointer,ParameterPieces::isthis));
  callspec.setParam(6,"this_plain",
      pieces(reg,0x30,plainPointer,ParameterPieces::isthis));

  Varnode *fspecPrimary = fd->newVarnodeCallSpecs(&callspec);
  Varnode *fspecAlias = fd->newVarnodeCallSpecs(&callspec);
  const uintb callspecEncoding = (uintb)(uintp)&callspec;
  Varnode *constantMimic = fd->newConstant(sizeof(&callspec),callspecEncoding);
  fd->opSetInput(callOp,fspecPrimary,0);
  fd->opSetInput(callOp,fd->newConstant(8,0x7180),1);
  fd->opSetInput(callOp,fd->newConstant(8,0x11223344),2);
  fd->opSetInput(callOp,fd->newConstant(4,0x55667788),3);
  fd->opSetInput(callOp,fd->newConstant(8,0),4);
  fd->opSetInput(callOp,fd->newConstant(8,0x99a8),5);
  fd->opSetInput(callOp,fd->newConstant(8,0xc1d8),6);
  fd->opSetInput(callOp,fd->newConstant(8,0xea40),7);
  fd->opSetInput(callOp,fd->newConstant(1,0x7f),8);

  cout << "fixture=TYPEOP-LOCALTYPE-DISPATCH-0001.call_input" << '\n';
  cout << "architecture=" << architecture.archid << '\n';
  cout << "representation.fspec_name=" << fspecPrimary->getSpace()->getName() << '\n';
  cout << "representation.fspec_type="
       << static_cast<int4>(fspecPrimary->getSpace()->getType()) << '\n';
  cout << "representation.constant_type="
       << static_cast<int4>(constantMimic->getSpace()->getType()) << '\n';
  emitBool("representation.primary_roundtrip",
      FuncCallSpecs::getFspecFromConst(fspecPrimary->getAddr()) == &callspec);
  emitBool("representation.alias_roundtrip",
      FuncCallSpecs::getFspecFromConst(fspecAlias->getAddr()) == &callspec);
  emitBool("representation.alias_distinct_varnode",fspecPrimary != fspecAlias);
  emitBool("representation.alias_same_address",fspecPrimary->getAddr() == fspecAlias->getAddr());
  emitBool("representation.constant_same_offset",
      constantMimic->getOffset() == fspecPrimary->getOffset());
  emitBool("representation.constant_not_fspec",
      constantMimic->getSpace()->getType() != IPTR_FSPEC);

  emitCase("slot0_target",callOp,0,
      factory->getBase(fspecPrimary->getSize(),TYPE_UNKNOWN),(ProtoParameter *)0,factory);
  emitCase("locked_ptr_equal",callOp,1,charPointer,callspec.getParam(0),factory);
  emitCase("locked_small_fits",callOp,2,int4Type,callspec.getParam(1),factory);
  emitCase("unlocked_param",callOp,3,
      factory->getBase(4,TYPE_UNKNOWN),callspec.getParam(2),factory);
  emitCase("locked_void",callOp,4,
      factory->getBase(8,TYPE_UNKNOWN),callspec.getParam(3),factory);
  emitCase("locked_oversize",callOp,5,
      factory->getBase(8,TYPE_UNKNOWN),callspec.getParam(4),factory);
  emitCase("unlocked_this_struct",callOp,6,objectPointer,callspec.getParam(5),factory);
  emitCase("unlocked_this_plain",callOp,7,
      factory->getBase(8,TYPE_UNKNOWN),callspec.getParam(6),factory);
  emitCase("missing_param",callOp,8,
      factory->getBase(1,TYPE_UNKNOWN),(ProtoParameter *)0,factory);

  callspec.getParam(0)->setTypeLock(false);
  emitCase("locked_ptr_after_unlock",callOp,1,
      factory->getBase(8,TYPE_UNKNOWN),callspec.getParam(0),factory);
  callspec.getParam(0)->setTypeLock(true);
  emitCase("locked_ptr_after_relock",callOp,1,charPointer,callspec.getParam(0),factory);

  fd->opSetInput(callOp,fspecAlias,0);
  emitCase("fspec_alias",callOp,1,charPointer,callspec.getParam(0),factory);
  fd->opSetInput(callOp,constantMimic,0);
  emitCase("constant_same_offset",callOp,1,
      factory->getBase(8,TYPE_UNKNOWN),callspec.getParam(0),factory);
}

} // namespace

int main(int argc,char **argv)

{
  try {
    if (argc != 3)
      throw invalid_argument("usage: typeop_local_type_1204 SPEC_DIRECTORY BINARY");
    runFixture(argv[1],argv[2]);
    return 0;
  }
  catch(const LowlevelError &err) {
    cerr << "LowlevelError: " << err.explain << '\n';
  }
  catch(const exception &err) {
    cerr << "std::exception: " << err.what() << '\n';
  }
  return 1;
}

File: tests/oracle/typeop_local_type_1204.rs
// TYPEOP-LOCALTYPE-DISPATCH-0001: current-Rugra comparand for the locked
// Ghidra TypeOpCall::getInputLocal fixture.
//
// This deliberately exercises the production TypeOpCall trait method.  The
// callspec is installed in Funcdata and represented by the same Iop/index
// annotation that Funcdata::new_varnode_call_specs currently produces.  The
// output is expected to differ from Ghidra 12.0.4: TypeOpCall has no
// get_input_local override and the legacy Varnode address-space enum has no
// FSPEC variant.  The fixture therefore reports None; it does not emulate the
// missing behavior in test code.

use std::sync::Arc;

use rugra::address::Address;
use rugra::fspec::{protoparam_flags, FuncCallSpecs, FuncProto, ProtoParameter};
use rugra::funcdata::Funcdata;
use rugra::op::{PcodeOp, PcodeOpRef};
use rugra::opcodes::OpCode;
use rugra::space::SpaceType;
use rugra::type_system::datatype::{Datatype, TypeMetatype};
use rugra::type_system::typefactory::{SizeArchInputs, TypeFactory};
use rugra::typeop::{TypeOp, TypeOpCall};

fn emit_bool(key: &str, value: bool) {
    println!("{key}={}", if value { 1 } else { 0 });
}

fn option_identity(actual: Option<&Arc<Datatype>>, expected: &Arc<Datatype>) -> bool {
    actual.is_some_and(|datatype| Arc::ptr_eq(datatype, expected))
}

fn repeat_identity(left: Option<&Arc<Datatype>>, right: Option<&Arc<Datatype>>) -> bool {
    match (left, right) {
        (Some(left), Some(right)) => Arc::ptr_eq(left, right),
        _ => false,
    }
}

fn emit_case(
    name: &str,
    typeop: &TypeOpCall,
    op: &PcodeOp,
    slot: usize,
    expected: &Arc<Datatype>,
    param: Option<&ProtoParameter>,
    fallback: &Arc<Datatype>,
) {
    let actual = typeop.get_input_local(op, slot);
    let repeat = typeop.get_input_local(op, slot);
    println!("case.{name}.slot={slot}");
    println!(
        "case.{name}.input_size={}",
        op.get_in(slot)
            .expect("fixture input slot")
            .read()
            .unwrap()
            .size
    );
    emit_bool(&format!("case.{name}.param_present"), param.is_some());
    emit_bool(
        &format!("case.{name}.param_locked"),
        param.is_some_and(ProtoParameter::is_type_locked),
    );
    emit_bool(
        &format!("case.{name}.param_this"),
        param.is_some_and(ProtoParameter::is_this_pointer),
    );
    match actual.as_ref() {
        Some(datatype) => {
            println!("case.{name}.result_type={}", datatype.print_raw());
            println!("case.{name}.result_meta={:?}", datatype.get_metatype());
            println!("case.{name}.result_size={}", datatype.get_size());
        }
        None => {
            println!("case.{name}.result_type=NONE");
            println!("case.{name}.result_meta=NONE");
            println!("case.{name}.result_size=NONE");
        }
    }
    emit_bool(
        &format!("case.{name}.expected_identity"),
        option_identity(actual.as_ref(), expected),
    );
    emit_bool(
        &format!("case.{name}.param_identity"),
        param.is_some_and(|parameter| option_identity(actual.as_ref(), &parameter.data_type)),
    );
    emit_bool(
        &format!("case.{name}.fallback_identity"),
        option_identity(actual.as_ref(), fallback),
    );
    emit_bool(
        &format!("case.{name}.repeat_identity"),
        repeat_identity(actual.as_ref(), repeat.as_ref()),
    );
}

fn callspec_identity(fd: &Funcdata, index: usize, op: &PcodeOpRef) -> bool {
    match (fd.get_call_specs(index), fd.get_call_specs_of_op(op)) {
        (Some(expected), Some(actual)) => std::ptr::eq(expected, actual),
        _ => false,
    }
}

fn main() {
    let mut factory = TypeFactory::new(8);
    factory.setup_sizes(&SizeArchInputs {
        stack_spacebase_size: Some(8),
        default_data_space_addr_size: 8,
        default_size: 8,
        far_pointer: None,
    });

    let void_type = factory.get_type_void();
    let char_type = factory
        .get_type_char_named("char")
        .expect("canonical char type");
    let char_pointer = factory.get_type_pointer(8, char_type, 1);
    let int4_type = factory
        .get_base(4, TypeMetatype::Int)
        .expect("canonical int4 type");
    let uint8_type = factory
        .get_base(8, TypeMetatype::Uint)
        .expect("canonical uint8 type");
    let oversize_type = factory.get_array(uint8_type, 2);
    let object_type = factory.create_struct("typeop_call_fixture_object");
    let object_pointer = factory.get_type_pointer(8, object_type, 1);
    let plain_pointer = factory.get_type_pointer(8, int4_type.clone(), 1);

    let mut prototype = FuncProto::new("typeop_call_local_fixture".to_string(), void_type.clone());
    let mut locked_ptr = ProtoParameter::new(
        "locked_ptr".to_string(),
        char_pointer.clone(),
        Address::new(0x00),
    );
    locked_ptr.flags |= protoparam_flags::TYPE_LOCKED;
    prototype.add_parameter(locked_ptr);
    let mut locked_small = ProtoParameter::new(
        "locked_small".to_string(),
        int4_type.clone(),
        Address::new(0x08),
    );
    locked_small.flags |= protoparam_flags::TYPE_LOCKED;
    prototype.add_parameter(locked_small);
    prototype.add_parameter(ProtoParameter::new(
        "unlocked_int".to_string(),
        int4_type.clone(),
        Address::new(0x10),
    ));
    let mut locked_void = ProtoParameter::new(
        "locked_void".to_string(),
        void_type.clone(),
        Address::new(0x18),
    );
    locked_void.flags |= protoparam_flags::TYPE_LOCKED;
    prototype.add_parameter(locked_void);
    let mut locked_oversize = ProtoParameter::new(
        "locked_oversize".to_string(),
        oversize_type.clone(),
        Address::new(0x20),
    );
    locked_oversize.flags |= protoparam_flags::TYPE_LOCKED;
    prototype.add_parameter(locked_oversize);
    let mut this_struct = ProtoParameter::new(
        "this_struct".to_string(),
        object_pointer.clone(),
        Address::new(0x28),
    );
    this_struct.flags |= protoparam_flags::THIS_POINTER;
    prototype.add_parameter(this_struct);
    let mut this_plain =
        ProtoParameter::new("this_plain".to_string(), plain_pointer, Address::new(0x30));
    this_plain.flags |= protoparam_flags::THIS_POINTER;
    prototype.add_parameter(this_plain);

    let mut fd = Funcdata::new("typeop_call_local_fixture", Address::new(0x500000), 0x100);
    let callspec_index = fd.add_call_specs(FuncCallSpecs::new(Address::new(0x500010), prototype));
    let fspec_primary = fd.new_varnode_call_specs(callspec_index);
    let fspec_alias = fd.new_varnode_call_specs(callspec_index);
    let constant_mimic = fd.new_constant(std::mem::size_of::<usize>(), callspec_index as u64);
    let input1 = fd.new_constant(8, 0x7180);
    let input2 = fd.new_constant(8, 0x11223344);
    let input3 = fd.new_constant(4, 0x55667788);
    let input4 = fd.new_constant(8, 0);
    let input5 = fd.new_constant(8, 0x99a8);
    let input6 = fd.new_constant(8, 0xc1d8);
    let input7 = fd.new_constant(8, 0xea40);
    let input8 = fd.new_constant(1, 0x7f);

    let op = fd.new_op(9, Address::new(0x500010));
    fd.op_set_opcode(&op, OpCode::CPUI_CALL);
    fd.op_set_input(&op, fspec_primary.clone(), 0);
    fd.op_set_input(&op, input1, 1);
    fd.op_set_input(&op, input2, 2);
    fd.op_set_input(&op, input3, 3);
    fd.op_set_input(&op, input4, 4);
    fd.op_set_input(&op, input5, 5);
    fd.op_set_input(&op, input6, 6);
    fd.op_set_input(&op, input7, 7);
    fd.op_set_input(&op, input8, 8);

    println!("fixture=TYPEOP-LOCALTYPE-DISPATCH-0001.call_input");
    println!("architecture=current-rust-size-configuration-only");
    println!("representation.fspec_name=iop");
    println!("representation.fspec_type={}", SpaceType::Iop as u32);
    println!(
        "representation.constant_type={}",
        SpaceType::Constant as u32
    );
    emit_bool(
        "representation.primary_roundtrip",
        callspec_identity(&fd, callspec_index, &op),
    );
    {
        let mut call = op.0.write().unwrap();
        call.inrefs[0] = fspec_alias.clone();
    }
    emit_bool(
        "representation.alias_roundtrip",
        callspec_identity(&fd, callspec_index, &op),
    );
    emit_bool(
        "representation.alias_distinct_varnode",
        !Arc::ptr_eq(&fspec_primary, &fspec_alias),
    );
    emit_bool(
        "representation.alias_same_address",
        fspec_primary.read().unwrap().get_space() == fspec_alias.read().unwrap().get_space()
            && fspec_primary.read().unwrap().get_offset()
                == fspec_alias.read().unwrap().get_offset(),
    );
    emit_bool(
        "representation.constant_same_offset",
        constant_mimic.read().unwrap().get_offset() == fspec_primary.read().unwrap().get_offset(),
    );
    emit_bool("representation.constant_not_fspec", true);
    {
        let mut call = op.0.write().unwrap();
        call.inrefs[0] = fspec_primary.clone();
    }

    let typeop = TypeOpCall;
    {
        let call = op.0.read().unwrap();
        let callspec = fd.get_call_specs(callspec_index).unwrap();
        emit_case(
            "slot0_target",
            &typeop,
            &call,
            0,
            &factory.get_base(8, TypeMetatype::Unknown).unwrap(),
            None,
            &factory.get_base(8, TypeMetatype::Unknown).unwrap(),
        );
        emit_case(
            "locked_ptr_equal",
            &typeop,
            &call,
            1,
            &char_pointer,
            callspec.prototype.get_param(0),
            &factory.get_base(8, TypeMetatype::Unknown).unwrap(),
        );
        emit_case(
            "locked_small_fits",
            &typeop,
            &call,
            2,
            &int4_type,
            callspec.prototype.get_param(1),
            &factory.get_base(8, TypeMetatype::Unknown).unwrap(),
        );
        emit_case(
            "unlocked_param",
            &typeop,
            &call,
            3,
            &factory.get_base(4, TypeMetatype::Unknown).unwrap(),
            callspec.prototype.get_param(2),
            &factory.get_base(4, TypeMetatype::Unknown).unwrap(),
        );
        emit_case(
            "locked_void",
            &typeop,
            &call,
            4,
            &factory.get_base(8, TypeMetatype::Unknown).unwrap(),
            callspec.prototype.get_param(3),
            &factory.get_base(8, TypeMetatype::Unknown).unwrap(),
        );
        emit_case(
            "locked_oversize",
            &typeop,
            &call,
            5,
            &factory.get_base(8, TypeMetatype::Unknown).unwrap(),
            callspec.prototype.get_param(4),
            &factory.get_base(8, TypeMetatype::Unknown).unwrap(),
        );
        emit_case(
            "unlocked_this_struct",
            &typeop,
            &call,
            6,
            &object_pointer,
            callspec.prototype.get_param(5),
            &factory.get_base(8, TypeMetatype::Unknown).unwrap(),
        );
        emit_case(
            "unlocked_this_plain",
            &typeop,
            &call,
            7,
            &factory.get_base(8, TypeMetatype::Unknown).unwrap(),
            callspec.prototype.get_param(6),
            &factory.get_base(8, TypeMetatype::Unknown).unwrap(),
        );
        emit_case(
            "missing_param",
            &typeop,
            &call,
            8,
            &factory.get_base(1, TypeMetatype::Unknown).unwrap(),
            None,
            &factory.get_base(1, TypeMetatype::Unknown).unwrap(),
        );
    }

    fd.get_call_specs_mut(callspec_index)
        .unwrap()
        .prototype
        .parameters[0]
        .flags &= !protoparam_flags::TYPE_LOCKED;
    {
        let call = op.0.read().unwrap();
        let callspec = fd.get_call_specs(callspec_index).unwrap();
        emit_case(
            "locked_ptr_after_unlock",
            &typeop,
            &call,
            1,
            &factory.get_base(8, TypeMetatype::Unknown).unwrap(),
            callspec.prototype.get_param(0),
            &factory.get_base(8, TypeMetatype::Unknown).unwrap(),
        );
    }
    fd.get_call_specs_mut(callspec_index)
        .unwrap()
        .prototype
        .parameters[0]
        .flags |= protoparam_flags::TYPE_LOCKED;
    {
        let call = op.0.read().unwrap();
        let callspec = fd.get_call_specs(callspec_index).unwrap();
        emit_case(
            "locked_ptr_after_relock",
            &typeop,
            &call,
            1,
            &char_pointer,
            callspec.prototype.get_param(0),
            &factory.get_base(8, TypeMetatype::Unknown).unwrap(),
        );
    }

    {
        let mut call = op.0.write().unwrap();
        call.inrefs[0] = fspec_alias;
    }
    {
        let call = op.0.read().unwrap();
        let callspec = fd.get_call_specs(callspec_index).unwrap();
        emit_case(
            "fspec_alias",
            &typeop,
            &call,
            1,
            &char_pointer,
            callspec.prototype.get_param(0),
            &factory.get_base(8, TypeMetatype::Unknown).unwrap(),
        );
    }
    {
        let mut call = op.0.write().unwrap();
        call.inrefs[0] = constant_mimic;
    }
    {
        let call = op.0.read().unwrap();
        let callspec = fd.get_call_specs(callspec_index).unwrap();
        emit_case(
            "constant_same_offset",
            &typeop,
            &call,
            1,
            &factory.get_base(8, TypeMetatype::Unknown).unwrap(),
            callspec.prototype.get_param(0),
            &factory.get_base(8, TypeMetatype::Unknown).unwrap(),
        );
    }
    println!("rugra.typeop_call_get_input_local.status=MISMATCH");
    println!("rugra.fspec_varnode_representation.status=MISMATCH");
}
